internal/services/v1/experimental.go

```diff
@@ -386,6 +406,153 @@ func (es *experimentalServer) BulkExportRelationships(
 	return nil
 }
 
+func (es *experimentalServer) BulkCheckPermission(ctx context.Context, req *v1.BulkCheckPermissionRequest) (*v1.BulkCheckPermissionResponse, error) {
+	atRevision, checkedAt, err := consistency.RevisionFromContext(ctx)
+	if err != nil {
+		return nil, es.rewriteError(ctx, err)
+	}
+
+	// Identify checks with same permission+subject over different resources and group them. This is doable because
+	// the dispatching system already internally supports this kind of batching for performance.
+	groupingParams := groupingParameters{
+		atRevision:           atRevision,
+		maxCaveatContextSize: es.maxCaveatContextSize,
+		maximumAPIDepth:      es.maximumAPIDepth,
+	}
+	groupedItems, err := groupItems(ctx, groupingParams, req.Items)
+	if err != nil {
+		return nil, es.rewriteError(ctx, err)
+	}
+
+	bulkResponseMutex := sync.Mutex{}
+	resp := &v1.BulkCheckPermissionResponse{CheckedAt: checkedAt}
```

Comment:
I can see pros and cons to extracting these into a struct and then having `FailGroup()`,`AppendResults()`,`Response()` methods.

I don't know if one is fundamentally more efficient (closure vs type).

Reply:
*shrug* I think it would be about as complex

Reply:
I took a peak at the assembly and I don't think the difference is huge. Let's go with this until a profiler tells us otherwise.